### PR TITLE
RISC-V support for the SHA512

### PR DESCRIPTION
**Pull request by marcfedorow**

### CODE REVIEW

**Comment by marcfedorow** [comment]
If anyone is interested in such type of patches, I will
- require a CLA from my organization
- add some similar patches to the other ciphers (if applicable).

Please let me know if this worth doing.

**Comment by paulidale** [comment]
I think it is worthwhile adding better RISC-V support.

**Comment by marcfedorow** [comment]
I requested CLAs via email to legal@opensslfoundation.org.
I will append other ciphers' support to this PR.

**Comment by mattcaswell** [comment]
Close/reopen to kick CLA bot

**Comment by paulidale** [comment]
> I will append other ciphers' support to this PR.

There is no real problem making them separate pull requests.  At least if they won't cause merge/rebase conflicts.
Separate small PRs tend to be easier to review.

**Comment by marcfedorow** [comment]
> > I will append other ciphers' support to this PR.
> 
> There is no real problem making them separate pull requests. At least if they won't cause merge/rebase conflicts.
> Separate small PRs tend to be easier to review.

If so, this one is ready for review

File: crypto/sha/sha512.c
Comment (paulidale) [inline crypto/sha/sha512.c:437]: Formatting nit: spaces either size of the equals.
This repeats several times throughout.

**Review by paulidale — APPROVED** [review]
This looks okay but my RISC-V assembly is non-existent so I'm trusting that part.

File: crypto/sha/sha512.c
Comment (marcfedorow) [reply crypto/sha/sha512.c:437]: I just copied formatting from existing code (e.g. see line 411).
Should I add spaces around the '='?

Comment (paulidale) [reply crypto/sha/sha512.c:437]: Yes please.

Comment (marcfedorow) [reply crypto/sha/sha512.c]: Force pushed tabulation fixes

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by marcfedorow** [comment]
I am very sorry for the late change but there are some issues with aliases operand order in updated riscv-p-spec.
The opcode/encoding for bit-wise pick is frozen but its name and order of sources differs (in terms of assembler but not machine code).
I replaced opcode name with machine encoding so there will be no need to fix any issues in case of renaming/reordering.
Once again sorry for the late change.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by marcfedorow** [comment]
@paulidale I see the `approval: review pending` tag.
Its hint says: `This pull request needs review by a committer`.
What should I do?

**Comment by paulidale** [comment]
You don't need to do anything.  Those tags are for us to track the state of the pull request.  As soon as you made the change, the reviews were invalidated and needed to be done again.  I changed the status to indicate this but forgot to remove the _ready to merge_ tag.  @t8m removed that but missed removing the _review pending_ tag.  It's all a fairly manual process.

A note: the reviewers find it easier if changes are done with _fixup_ commits rather than force pushes:
```sh
$ git add ...
$ git commit --fixup <commit ID>
$ git push
```
This lets us see the changes in isolation.  For small/trivial things, it speeds up reviewing considerable since we often don't need to look at the bulk of the changes again.  _Fixup_ commits are also squashed automatically during the merge process.

**Comment by paulidale** [comment]
Merged to master.  Thanks for the contribution.

**Comment by ZenithalHourlyRate** [comment]
Hi, I've tried to benchmark the RV64 part of this PR and #16710 using [Zk/Zb enabled](https://github.com/chipsalliance/rocket-chip/pull/2950) rocket chip running on a FPGA (detailed setup [here](https://github.com/chipsalliance/rocket-chip/pull/2950)). The results are remarkable.
```
# C version
sha256             136.14k      376.49k      754.43k     1010.35k     1144.15k     1159.40k
sha512             135.83k      539.50k     1092.58k     1764.09k     2135.38k     2184.53k
# inline asm version
sha256             200.00k      594.04k     1282.90k     1813.40k     2059.07k     2104.17k
sha512             243.83k      977.74k     2052.78k     4004.56k     5382.14k     5444.95k
```

However, there are two issues about compiling.
1. The code path is not included when `./config linux64-riscv64` (note, no `no-asm` option) because the macro `OPENSSL_NO_ASM` is defined. This is caused by `no-asm` since there is no `asm_arch` defined. Thus to make this PR effective I need to cherry-pick "Add riscv64 asm_arch to linux64-riscv64 target" from #17640
```
# perl configdata.pm --dump
Disabled features:
    asm                 [no asm_arch]    OPENSSL_NO_ASM
```
2. About the `#if __riscv_zknh` condition, the `__riscv_zknh` macro is actually defined with I `CC="riscv64-unknown-linux-musl-gcc -march=rv64gc_zknh_zbkb" ./config linux64-riscv64` but its value is 0. (Surprisingly, when I compile with `-march="rv64gc_zknh1_zbkb1"`, `__riscv_zknh` would be `1000000`). This is actually still a WIP in the toolchain (ref to https://github.com/riscv-non-isa/riscv-c-api-doc/pull/23). I think a better way it to test by `#if defined(__riscv_zknh)`

**Comment by paulidale** [comment]
@[ZenithalHourlyRate](https://github.com/ZenithalHourlyRate), since you've come up with solutions to this, is there a chance you could submit a pull request fixing things?

**Comment by ZenithalHourlyRate** [comment]
> @[ZenithalHourlyRate](https://github.com/ZenithalHourlyRate), since you've come up with solutions to this, is there a chance you could submit a pull request fixing things?

@paulidale I've sent a PR #18275

**Comment by ZenithalHourlyRate** [comment]
> Hi, I've tried to benchmark the RV64 part of this PR and #16710 using [Zk/Zb enabled](https://github.com/chipsalliance/rocket-chip/pull/2950) rocket chip running on a FPGA (detailed setup [here](https://github.com/chipsalliance/rocket-chip/pull/2950)). The results are remarkable.

the benchmark of the RV32 part
```
# C version
sha256             173.80k      439.57k      825.62k     1061.67k     1165.99k     1174.19k
sha512              44.97k      183.20k      316.16k      519.17k      655.36k      666.28k
# inline asm version
sha256             269.40k      721.62k     1438.91k     1909.42k     2088.96k     2113.54k
sha512             121.69k      494.76k      909.91k     1414.08k     1671.17k     1703.94k
```

**Comment by marcfedorow** [comment]
@ZenithalHourlyRate
I just compiled this

```
int main()
{
    printf("zkn  = %d\n", __riscv_zkn );
    printf("zkne = %d\n", __riscv_zkne);
    printf("zknd = %d\n", __riscv_zknd);
    printf("zknh = %d\n", __riscv_zknh);
    return 0;
}

```
with `-march=rv64izk` with gcc 11.2 (with my patches) and gcc 12.1 (stable, no additional patches).

This is the output;
```
zkn  = 1000000
zkne = 1000000
zknd = 1000000
zknh = 1000000
```
(Where 1 is major version, 000 is minor version, 000 is patch version).

In the file `riscv-gcc/gcc/common/config/riscv/riscv-common.c` (gcc11) or riscv-gcc/gcc/common/config/riscv/riscv-common.cc (gcc12)
you should have a table
```
static const riscv_implied_info_t riscv_implied_info[] =
```
which should contain the following:
```
  {"zk", "zkn"},
  {"zk", "zkr"},
  {"zk", "zkt"},
  {"zkn", "zbkb"},
  {"zkn", "zbkc"},
  {"zkn", "zbkx"},
  {"zkn", "zkne"},
  {"zkn", "zknd"},
  {"zkn", "zknh"},
  {"zks", "zbkb"},
  {"zks", "zbkc"},
  {"zks", "zbkx"},
  {"zks", "zksed"},
  {"zks", "zksh"},
```
and a table
```
static const struct riscv_ext_version riscv_ext_version_table[] =
```
should have
```
  {"zbkb",  ISA_SPEC_CLASS_NONE, 1, 0},
  {"zbkc",  ISA_SPEC_CLASS_NONE, 1, 0},
  {"zbkx",  ISA_SPEC_CLASS_NONE, 1, 0},
  {"zkne",  ISA_SPEC_CLASS_NONE, 1, 0},
  {"zknd",  ISA_SPEC_CLASS_NONE, 1, 0},
  {"zknh",  ISA_SPEC_CLASS_NONE, 1, 0},
  {"zkr",   ISA_SPEC_CLASS_NONE, 1, 0},
  {"zksed", ISA_SPEC_CLASS_NONE, 1, 0},
  {"zksh",  ISA_SPEC_CLASS_NONE, 1, 0},
  {"zkt",   ISA_SPEC_CLASS_NONE, 1, 0},

  {"zk",    ISA_SPEC_CLASS_NONE, 1, 0},
  {"zkn",   ISA_SPEC_CLASS_NONE, 1, 0},
  {"zks",   ISA_SPEC_CLASS_NONE, 1, 0},
```

This mechanism is presented for compatibility.

**Pull request closed without merging**
